Kernel32LibraryTests creates JNA proxy classes

With the current tests policy in place, this hits SecurityException (e.g. http://build-us-00.elasticsearch.org/job/es_core_master_window-2012/1004/console).

Does the functionality or just the test need to create these proxies? Can we avoid it? Otherwise, we can add this permission, but it is a little bit strange that JNA for e.g. unix mlockall does not need to create proxies but on windows it does.

 Caused by: java.security.AccessControlException: access denied ("java.lang.reflect.ReflectPermission" "newProxyInPackage.org.elasticsearch.common.jna")
  1>     java.security.AccessControlContext.checkPermission(AccessControlContext.java:457)
  1>     java.security.AccessController.checkPermission(AccessController.java:884)
  1>     java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
  1>     [...java.lang.reflect.*]
  1>     com.sun.jna.Native.loadLibrary(Native.java:415)
  1>     com.sun.jna.Native.loadLibrary(Native.java:391)
  1>     org.elasticsearch.common.jna.Kernel32Library.<init>(Kernel32Library.java:50)
  1>     org.elasticsearch.common.jna.Kernel32Library.<init>(Kernel32Library.java:36)
  1>     org.elasticsearch.common.jna.Kernel32Library$Holder.<clinit>(Kernel32Library.java:45)
  1>     org.elasticsearch.common.jna.Kernel32Library.getInstance(Kernel32Library.java:60)
  1>     org.elasticsearch.common.jna.Kernel32LibraryTests.testKernel32Library(Kernel32LibraryTests.java:74)

@rmuir thanks for spotting it.

> Does the functionality or just the test need to create these proxies? Can we avoid it? Otherwise, we can add this permission, but it is a little bit strange that JNA for e.g. unix mlockall does not need to create proxies but on windows it does.

The functionality makes a native call on Windows to add a console control handler. The current implementation uses proxification which is the best documented way to make a native call on Windows with the JNA library. The other way is to use [Direct Mapping](https://github.com/twall/jna/blob/master/www/DirectMapping.md) like what is done in the unix/mlockall function.

If these proxy creation is really annoying, I think I can change the code to use Direct Mapping in Kernel32Library too. 

Besides this, I didn't see a test for the `Native.tryMlockall()` method. I added one locally and it fails with SecurityManager enabled and the current test policy: 
```
Caused by: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "createSecurityManager")
```

> I really wonder if the native library we load here is worth the trouble?

#8993 relies on the [SetConsoleCtrlHandler function](https://msdn.microsoft.com/en-us/library/windows/desktop/ms683242(v=vs.85).aspx) and was added to ensure that a node is correctly shutdown on Windows when the user closes the console that starts Elasticsearch (either by clicking Close on the console window's window menu, or by clicking the End Task button command from Task Manager, or with a TASKILL command like a stop script would do). I really think that's very important to gracefully stop the node... Without this, it's like kill -9 the JVM.

So I suggest to keep the native library (it will also be used for the mlockall equivalent on Windows) and unify the way the native calls are made in CLibrary/Kernel32Libray and limit the permissions to grant in the policy file, and finally add a test for Native.tryMlockall().

--- a/dev-tools/tests.policy
+++ b/dev-tools/tests.policy
@@ -70,7 +70,9 @@ grant {
   // needed by groovy scripting
   permission java.lang.RuntimePermission "getProtectionDomain";
 
+  // needed for natives calls
   permission java.lang.RuntimePermission "loadLibrary.*";
+  permission java.lang.RuntimePermission "createSecurityManager";
 
   // reflection hacks:
   // needed for Striped64 (what is this doing), also enables unmap hack

--- a/src/main/java/org/elasticsearch/common/jna/Kernel32Library.java
+++ b/src/main/java/org/elasticsearch/common/jna/Kernel32Library.java
@@ -20,7 +20,6 @@
 package org.elasticsearch.common.jna;
 
 import com.google.common.collect.ImmutableList;
-import com.sun.jna.Library;
 import com.sun.jna.Native;
 import com.sun.jna.win32.StdCallLibrary;
 import org.elasticsearch.common.logging.ESLogger;
@@ -37,17 +36,18 @@ public class Kernel32Library {
 
     private static ESLogger logger = Loggers.getLogger(Kernel32Library.class);
 
-    private Kernel32 internal;
-
+    // Callbacks must be kept around in order to be able to be called later,
+    // when the Windows ConsoleCtrlHandler sends an event.
     private List<NativeHandlerCallback> callbacks = new ArrayList<>();
 
+    // Native library instance must be kept around for the same reason.
     private final static class Holder {
         private final static Kernel32Library instance = new Kernel32Library();
     }
 
     private Kernel32Library() {
         try {
-            internal = (Kernel32)Native.synchronizedLibrary((Kernel32)Native.loadLibrary("kernel32", Kernel32.class));
+            Native.register("kernel32");
             logger.debug("windows/Kernel32 library loaded");
         } catch (NoClassDefFoundError e) {
             logger.warn("JNA not found. native methods and handlers will be disabled.");
@@ -60,14 +60,19 @@ public class Kernel32Library {
         return Holder.instance;
     }
 
+    /**
+     * Adds a Console Ctrl Handler.
+     *
+     * @param handler
+     * @return true if the handler is correctly set
+     * @throws java.lang.UnsatisfiedLinkError if the Kernel32 library is not loaded or if the native function is not found
+     * @throws java.lang.NoClassDefFoundError if the library for native calls is missing
+     */
     public boolean addConsoleCtrlHandler(ConsoleCtrlHandler handler) {
-        if (internal == null) {
-            throw new UnsupportedOperationException("windows/Kernel32 library not loaded, console ctrl handler cannot be set");
-        }
         boolean result = false;
         if (handler != null) {
             NativeHandlerCallback callback = new NativeHandlerCallback(handler);
-            result = internal.SetConsoleCtrlHandler(callback, true);
+            result = SetConsoleCtrlHandler(callback, true);
             if (result) {
                 callbacks.add(callback);
             }
@@ -79,17 +84,16 @@ public class Kernel32Library {
         return ImmutableList.builder().addAll(callbacks).build();
     }
 
-    interface Kernel32 extends Library {
-
-        /**
-         * Registers a Console Ctrl Handler.
-         *
-         * @param handler
-         * @param add
-         * @return true if the handler is correctly set
-         */
-        public boolean SetConsoleCtrlHandler(StdCallLibrary.StdCallCallback handler, boolean add);
-    }
+    /**
+     * Native call to the Kernel32 API to set a new Console Ctrl Handler.
+     *
+     * @param handler
+     * @param add
+     * @return true if the handler is correctly set
+     * @throws java.lang.UnsatisfiedLinkError if the Kernel32 library is not loaded or if the native function is not found
+     * @throws java.lang.NoClassDefFoundError if the library for native calls is missing
+     */
+    public native boolean SetConsoleCtrlHandler(StdCallLibrary.StdCallCallback handler, boolean add);
 
     /**
      * Handles consoles event with WIN API

--- a/src/test/java/org/elasticsearch/common/jna/NativesTests.java
+++ b/src/test/java/org/elasticsearch/common/jna/NativesTests.java
@@ -20,7 +20,6 @@
 package org.elasticsearch.common.jna;
 
 import org.apache.lucene.util.Constants;
-import org.apache.lucene.util.LuceneTestCase.AwaitsFix;
 import org.elasticsearch.common.jna.Kernel32Library.ConsoleCtrlHandler;
 import org.elasticsearch.test.ElasticsearchTestCase;
 import org.junit.After;
@@ -32,8 +31,7 @@ import java.util.Map;
 
 import static org.hamcrest.Matchers.equalTo;
 
-@AwaitsFix(bugUrl = "https://github.com/elasticsearch/elasticsearch/issues/9802")
-public class Kernel32LibraryTests extends ElasticsearchTestCase {
+public class NativesTests extends ElasticsearchTestCase {
 
     /**
      * Those properties are set by the JNA Api and if not ignored,
@@ -65,7 +63,16 @@ public class Kernel32LibraryTests extends ElasticsearchTestCase {
     }
 
     @Test
-    public void testKernel32Library() {
+    public void testTryMlockall() {
+        Natives.tryMlockall();
+
+        if (Constants.WINDOWS) {
+            assertFalse("Memory locking is not available on Windows platforms", Natives.LOCAL_MLOCKALL);
+        }
+    }
+
+    @Test
+    public void testAddConsoleCtrlHandler() {
         ConsoleCtrlHandler handler = new ConsoleCtrlHandler() {
             @Override
             public boolean handle(int code) {
@@ -73,18 +80,21 @@ public class Kernel32LibraryTests extends ElasticsearchTestCase {
             }
         };
 
-        assertNotNull(Kernel32Library.getInstance());
-        assertThat(Kernel32Library.getInstance().getCallbacks().size(), equalTo(0));
+        Natives.addConsoleCtrlHandler(handler);
 
         if (Constants.WINDOWS) {
-            assertTrue(Kernel32Library.getInstance().addConsoleCtrlHandler(handler));
+            assertNotNull(Kernel32Library.getInstance());
             assertThat(Kernel32Library.getInstance().getCallbacks().size(), equalTo(1));
+
         } else {
+            assertNotNull(Kernel32Library.getInstance());
+            assertThat(Kernel32Library.getInstance().getCallbacks().size(), equalTo(0));
+
             try {
                 Kernel32Library.getInstance().addConsoleCtrlHandler(handler);
                 fail("should have thrown an unsupported operation exception");
-            } catch (UnsupportedOperationException e) {
-                assertThat(e.getMessage(), e.getMessage().contains("console ctrl handler cannot be set"), equalTo(true));
+            } catch (UnsatisfiedLinkError e) {
+                // UnsatisfiedLinkError is expected
             }
         }
     }

